kotlinx-coroutines-core/common/src/internal/LimitedDispatcher.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+import kotlin.jvm.*
+
+/**
+ * The result of .limitedParallelism(x) call, dispatcher
+ * that wraps the given dispatcher, but limits the parallelism level, while
+ * trying to emulate fairness.
+ */
+internal class LimitedDispatcher(
+    private val dispatcher: CoroutineDispatcher,
+    private val parallelism: Int
+) : CoroutineDispatcher(), Runnable, Delay by (dispatcher as? Delay ?: DefaultDelay) {
```

Comment:
does it make sense to override `limitedParallelism` for this class too? smth like
```
    override fun limitedParallelism(parallelism: Int): CoroutineDispatcher {
        parallelism.checkParallelism()
        return if (parallelism >= this.parallelism) this else LimitedDispatcher(this, parallelism)
    }
``` 
to avoid overhead when stacking multiple limitedParalelism calls

Reply:
It could also just wrap the underlying dispatcher directly in that case, instead of wrapping another instance of `LimitedDispatcher`.

Reply:
no, in this case you may fully load underlaying dispatcher (just because these limited dispatchers will schedule their workers to different threads of underlying dispatcher bypassing limited parallelism mechanism)